nat-lab/guidelines.md

```diff
@@ -0,0 +1,215 @@
+   - Tests should leave no residual configurations even in case of failure.
+
+   **Why:**
+   - Some tests tinker with iptables / routing tables and some other significant stuff, which might lead to failure of other tests if not properly cleaned.
+
+   **Example**:
+   - DON'T:
+   ```
+   def test_1():
+      # for the purpose of example, break_tcp_conn_to_host is not context manager and doesnt have a proper cleanup
+
+      await connect_to(DERP_1)
+
+      async with asyncio_util.run_async_context(wait_for_event_peer(State.Disconnected)) as event:
+            await asyncio.wait_for(
+                asyncio.gather(
+                    *[
+                        break_tcp_conn_to_host(DERP_1)
+                        event,
+                    ]
+                ),
+                timeout
+            )
+
+   # this test will fail, because previous test didn't cleanup properly
+   def test_2():
+      async with asyncio_util.run_async_context(wait_for_event_peer(State.Connected)) as event:
+            await asyncio.wait_for(
+                asyncio.gather(
+                    *[
+                        connect_to(DERP_1)
+                        event,
+                    ]
+                ),
+                timeout
+            )
+   ```
+
+   - DO:
+   ```
+   def test():
+      # for the purpose of example, break_tcp_conn_to_host is not context manager and doesnt have a proper cleanup
+
+      await connect_to(DERP_1)
+
+      async with asyncio_util.run_async_context(wait_for_event_peer(State.Disconnected)) as event:
+            await asyncio.wait_for(
+                asyncio.gather(
+                    *[
+                        break_tcp_conn_to_host(DERP_1)
+                        event,
+                    ]
+                ),
+                timeout
+            )
+
+      restore_tcp_conn_to_host(DERP_1)
+   ```
+
+7. **Confirmation Beyond stdout:**
+   - Do not rely solely on stdout; explore alternative methods to confirm states using external tools.
+
+   **Why:**:
+   - Stdout might say "Connected", but there are no actual connection in conntrack records.
+
+   **Example:**
+   - Use connection tracker. Providing configuration to environment setup will enable it. When test being teardown, it will be checked, wether actual connection to vpn server was made.
+   ```
+   def test_connection():
+      env = setup_env(SetupParams(connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1, connection_tracker_config=generate_connection_tracker_config(vpn_1_limits=ConnectionLimits(1, 1))))
+      client.connect_to_vpn()
+      client.wait_for_peer_state(Connected)
+      assert "Connected" in client.get_stdout()
+   ```
+
+8. **Comprehensive Node State Checking:**
+   - Examine the states of all related nodes, not just a single one.
+
+   **Why:**
+   - By checking only one node, it is not made sure, that other node will be the same.
+
+   **Example:**  
+   To check if two nodes are connected via meshnet PING is used.
```

Comment:
This could be automated with a function that takes both nodes and verifies ping in both directions instead of asking devs to remember to do this.

Reply:
Ofc it could, but as if now, it is just a guideline. In the future, code quality will be improved accordingly and such function might appear.

Reply:
I understand that this is just an example but the positive example can be used to show the reader that the correct approach is to use higher level abstraction (like `verify_connectivity_between_nodes([alpha, beta])`) instead of teaching that it's fine to copy&paste snippets of code (even if they do what they should be doing).

Reply:
Updated